manifest-updater/manifest/storage.go

```diff
@@ -0,0 +1,260 @@
+	reader, err := s.client.Bucket(s.manifestBucketLocation).
+		Object(path).
+		NewReader(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get GCS object reader: %w", err)
+	}
+
+	return reader, nil
+}
+
+// S3Storage is a Storage that stores manifests in an S3 bucket
+type S3Storage struct {
+	client         *s3.S3
+	manifestBucket string
+}
+
+// newS3 creates an S3Writer that stores manifests in the specified S3 bucket
+func newS3(manifestBucket, region, profile string) (*S3Storage, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("making AWS session: %w", err)
+	}
+
+	config := aws.
+		NewConfig().
+		WithRegion(region).
+		WithCredentials(credentials.NewSharedCredentials("", profile))
+	s3Client := s3.New(sess, config)
+
+	return &S3Storage{
+		client:         s3Client,
+		manifestBucket: manifestBucket,
+	}, nil
+}
+
+func (s *S3Storage) WriteIngestorGlobalManifest(manifest IngestorGlobalManifest) error {
+	return s.writeManifest(manifest, "global-manifest.json")
+}
+
+func (s *S3Storage) WriteDataShareProcessorSpecificManifest(manifest DataShareProcessorSpecificManifest, dataShareProcessorName string) error {
+	return s.writeManifest(manifest, fmt.Sprintf("%s-manifest.json"))
+}
+
+func (s *S3Storage) writeManifest(manifest interface{}, path string) error {
+	log.Info().
+		Str("path", path).
+		Str("bucket", s.manifestBucket).
+		Msg("writing a manifest file")
+
+	jsonManifest, err := json.Marshal(manifest)
+	if err != nil {
+		return fmt.Errorf("failed to marshal manifest to JSON: %w", err)
+	}
+
+	input := &s3.PutObjectInput{
+		ACL:          aws.String(s3.BucketCannedACLPublicRead),
+		Body:         aws.ReadSeekCloser(bytes.NewReader(jsonManifest)),
+		Bucket:       aws.String(s.manifestBucket),
+		Key:          aws.String(path),
+		CacheControl: aws.String("no-cache"),
+		ContentType:  aws.String("application/json; charset=UTF-8"),
+	}
+
+	if _, err := s.client.PutObject(input); err != nil {
+		return fmt.Errorf("storage.PutObject: %w", err)
+	}
+
+	return nil
+}
+
+func (s *S3Storage) FetchDataShareProcessorSpecificManifest(dataShareProcessorName string) (*DataShareProcessorSpecificManifest, error) {
+	reader, err := s.getReader(fmt.Sprintf("%s-manifest.json", dataShareProcessorName))
+	if err != nil {
+		return nil, err
+	}
+
+	var manifest DataShareProcessorSpecificManifest
+	if err := json.NewDecoder(reader).Decode(&manifest); err != nil {
+		return nil, fmt.Errorf("error parsing manifest: %w", err)
+	}
+
+	return &manifest, nil
+}
+
+func (s *S3Storage) IngestorGlobalManifestExists() (bool, error) {
+	reader, err := s.getReader("global-manifest.json")
+	if err != nil {
+		return false, err
+	}
+
+	var manifest IngestorGlobalManifest
+	if err := json.NewDecoder(reader).Decode(&manifest); err != nil {
+		return false, fmt.Errorf("error parsing manifest: %w", err)
+	}
+
+	return true, nil
+}
+
+func (s *S3Storage) getReader(path string) (io.ReadCloser, error) {
```

Comment:
Add a comment to this method that the returned Body, being an io.ReadCloser, needs to be closed to avoid leaking.

Also, we then need to Close it everywhere this gets called.  json.Decoder doesn't, as it only needs a Reader, so it definitely doesn't know to Close on completion.

This probably should have been caught via a closer lint; we might need to check our CI linting.

Reply:
Good catch, I'm spoiled by Rust's automagic closing of readers on drop. The same applies to the `cloud.google.com/go/storage.Reader`s returned by `GCSStorage.getReader` which implements `io.ReadCloser`, too.